core_impl/src/main/java/com/google/instrumentation/trace/SpanImpl.java

```diff
@@ -0,0 +1,170 @@
+
+  static SpanImpl startSpan(SpanContext context,
+      @Nullable EnumSet<Options> options,
+      String name,
+      @Nullable SpanId parentSpanId,
+      StartEndHandler startEndHandler,
+      @Nullable TimestampConverter timestampConverter,
+      Clock clock) {
+    SpanImpl span = new SpanImpl(context, options, name, parentSpanId, startEndHandler, clock);
+    span.start(timestampConverter);
+    return span;
+  }
+
+  /**
+   * Returns the name of the {@code Span}.
+   *
+   * @return the name of the {@code Span}.
+   */
+  String getName() {
+    return name;
+  }
+
+  /**
+   * Returns an immutable representation of all the data from this {@code Span}.
+   *
+   * @return an immutable representation of all the data from this {@code Span}.
+   * @throws IllegalStateException if the Span doesn't have RECORD_EVENTS option.
+   */
+  SpanData toSpanData() {
+    checkState(
+        getOptions().contains(Options.RECORD_EVENTS),
+        "Getting SpanData for a Span without RECORD_EVENTS options.");
+    synchronized (this) {
+      return new SpanData(
+          getContext(),
+          parentSpanId,
+          name,
+          timestampConverter.convertNanoTime(startNanoTime),
+          Collections.<String, AttributeValue>emptyMap(),
+          Collections.<TimedEvent<Annotation>>emptyList(),
+          Collections.<TimedEvent<NetworkEvent>>emptyList(),
+          Collections.<Link>emptyList(),
+          hasBeenEnded ? status : null,
+          hasBeenEnded ? timestampConverter.convertNanoTime(endNanoTime) : null);
+    }
+  }
+
+  @Override
+  public void addAttributes(Map<String, AttributeValue> attributes) {
+    // TODO(bdrutu): Implement this.
+  }
+
+  @Override
+  public void addAnnotation(String description, Map<String, AttributeValue> attributes) {
+    // TODO(bdrutu): Implement this.
+  }
+
+  @Override
+  public void addAnnotation(Annotation annotation) {
+    // TODO(bdrutu): Implement this.
+  }
+
+  @Override
+  public void addNetworkEvent(NetworkEvent networkEvent) {
+    // TODO(bdrutu): Implement this.
+  }
+
+  @Override
+  public void addLink(Link link) {
+    // TODO(bdrutu): Implement this.
+  }
+
+  @Override
+  public void end(EndSpanOptions options) {
+    if (!getOptions().contains(Options.RECORD_EVENTS)) {
+      return;
+    }
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling end() on an ended Span.");
+        return;
+      }
+      status = options.getStatus();
+      endNanoTime = clock.nowNanos();
+      startEndHandler.onEnd(this);
+      hasBeenEnded = true;
+    }
+  }
+
+  /**
+   * Abstract class to handle the start and end operations for a {@link Span} only when the {@code
+   * Span} has {@link Span.Options#RECORD_EVENTS} option.
+   *
+   * <p>Implementation must avoid high overhead work in any of the methods because the code is
+   * executed on the critical path.
+   */
+  @ThreadSafe
+  abstract static class StartEndHandler {
```

Comment:
Done. Actually one instance of the handler can be (it will be) shared by multiple Spans at the same time so multiple threads can call into OpStart/OpEnd at the same time.